Public/Src/Sandbox/MacOs/Interop/Posix/memory.c

```diff
@@ -42,15 +42,58 @@ int GetRamUsageInfo(RamUsageInfo *buffer, long bufferSize)
     return KERN_SUCCESS;
 }
 
+static uint64_t ProcessTreeWorkingSetSize(pid_t pid, const rlim_t max_proc_count, bool *success)
+{
+    uint64_t children_usage = 0;
+
+    pid_t pids[max_proc_count];
+    int child_count = proc_listchildpids(pid, pids, (int) max_proc_count);
+    *success &= child_count >= 0;
+
+    for (int i = 0; (i < child_count) && *success; i++)
+    {
+        int pid = pids[i];
+        rusage_info_current rusage;
+        *success &= proc_pid_rusage(pid, RUSAGE_INFO_CURRENT, (void **)&rusage) == 0;
+        children_usage += rusage.ri_resident_size;
+
+        pid_t childs_pids[max_proc_count];
+        int count = proc_listchildpids(pid, childs_pids, (int) max_proc_count);
+        *success &= count >= 0;
+
+        if (count != 0)
+        {
+            children_usage += ProcessTreeWorkingSetSize(pid, max_proc_count, success);
+        }
+    }
+
+    return children_usage;
+
+}
+
 int GetPeakWorkingSetSize(pid_t pid, uint64_t *buffer)
```

Comment:
>GetPeakWorkingSetSize [](start = 4, length = 21)

getrusage's man page makes it look like you can call it twice for RUSAGE_SELF and RUSAGE_CHILDREN and then add the rs_maxrss field from both to get the overall peak. Can you do that to simplify all of the logic around querying the child processes individually?

Reply:
`getrusage` does not allow for pid queries, it is always relative to the calling process / thread. there is no other way to query the whole tree for its memory consumption.

Reply:
Ah. That's too bad. I guess to use the nice API we'd then have to first launch some special process to do reporting that would then spawn the pip. That'd be too much complexity and overhead. Might be worth adding a comment why getrusage isn't appropriate